### DISCUSSION TITLE
Hi curl team, is there any situation one curl instance can have multiple sockets for http communication?

### DISCUSSION THREAD

**Discussion by TheBetterSolution** [curl/curl#14951] [Q&A]
Hi curl team, is there any situation one curl instance can have multiple sockets?
In old curl version(before curl 7.44.0), I first record the socket for the curl instance in the SOCKETOPENFUNCTION, and I compare the input parameter: "socket" in  the SOCKETFUCNTION callback and the recorded socket, I found that are different if the DNS server isn't reachable:
  assert(curlRequest->socket != socket); //This assert will be triggered if the DNS server isn't reachable.

So, there are multiple socket(maybe one for DNS, one for http) for the same curl instance, my question is: is that possible there are  multiple sockets for http in one curl instance?

Thanks very much.

**Comment by icing** [comment]
There can be many sockets involved in a curl transfer, depending on protocol and other settings. A transfer might involved a DNS over HTTP lookup. A transfer might use "Happy Eyeballing" and try several IP addresses in parallel. A transfer might try HTTP/2 and HTTP/3 simultaneously.

Therefore, your `CURLMOPT_SOCKETFUNCTION` should be prepared to get all kinds of sockets, not directly connected to the final server connection.
